File: c-prog-using-ps_rss_vsz/hw1p1.py
"""hw1p1.py."""


import subprocess
import shlex
import os


def get_rss():
    """Find resident set size under /proc/ ."""
    p1 = subprocess.Popen(["ls", "/proc"],
                          stdout=subprocess.PIPE)
    p2 = subprocess.Popen(shlex.split("grep '[0-9]'"),
                          stdin=p1.stdout,
                          stdout=subprocess.PIPE,
                          stderr=subprocess.PIPE)
    p1.stdout.close()
    out, err = p2.communicate()
    # print('out: {0}', format(out))
    # print('err: {0}', format(err))
    # print(out.decode("utf-8"))
    res = out.decode("utf-8").strip().split('\n')
    # decode converts bytes to str

    for pid in res:
        parmList = ["ps", "-p", pid, "-o", "rss", ""]
        # os.execvp("/bin/ps", parmList)


if __name__ == "__main__":
    get_rss()

File: c-prog-using-ps_rss_vsz/hw1p1.c
/*
The declaration of execvpe is only included if _GNU_SOURCE is
defined before including unistd.h as it is a GNU extension.
*/
#define _GNU_SOURCE
#include <unistd.h>
#include<stdio.h>
#include<stdlib.h>
#include<dirent.h>

DIR *opendir(const char *name);

struct dirent *readdir(DIR *dirp);

void main()
{
  FILE *in;
  extern FILE *popen();
  char buff[512];
  char ch;
  char pid[20];
  pid_t pid;

  in = popen("ls /proc | grep '[0-9]'", "r");
  fgets(buff, 512, in);

  //printf("%s\n",buff );

  ch = fgetc(in);
  /*
  sprintf(pid ,"%d", ch);
  char *const parmList[6] = {"ps","-p",pid,"-o","rss",NULL};
  printf("%s",pid);
  execvp("/bin/ps", parmList);
  */

  while (ch != EOF)
  {
      printf ("%c", ch);
      sprintf(pid ,"%d", ch);
      char *const parmList[6] = {"ps","-p",pid,"-o","rss",NULL};
      execvp("/bin/ps", parmList);
      ch = fgetc(in);
  }


//  fclose(in);

  if (pclose(in) != 0)
        fprintf(stderr," Error: Failed to close command stream \n");



}
